activesupport/lib/active_support/core_ext/hash/build_nested.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class Hash
+  # Builds a new hash that's easy to merge with deep_merge
+  #
+  #   h1 = Hash.build_nested(:conferences, :tracks, :sessions, :keynote, presenter: '@tenderlove', topic: 'hotwire')
+  #   # => { conferences: { tracks: { sessions: { keynote: { presenter: "@tenderlove", :topic=>"hotwire" } } } } }
+  #
+  #   h1.deep_merge(h2) # => { a: false, b: { c: [1, 2, 3], x: [3, 4, 5] } }
+  #
+  # use build_nested with deep_merge to make modifications deep in the hash structure:
+  #
+  #   h1 = { a: { b: { c: { c1: 100 } } } }
+  #   h1.deep_merge(Hash.build_nested(:a, :b, :c, c2: 200, c3: 300))
+  #   # => { a: { b: { c: { c1: 100, c2: 200, c3: 300 } } } }
+  def self.build_nested(*args, **kwargs)
+    args.reverse.reduce(Hash.new.merge(**kwargs)) do |hash, arg|
+      { arg => hash }
+    end
+  end
```

Comment:
This seems to overlap somewhat with Ruby's [proposed `#bury` method](https://bugs.ruby-lang.org/issues/11747), which was [rejected](https://bugs.ruby-lang.org/issues/11747#note-4).  Though perhaps defining it as a class method instead of an instance method will address the concern (by making it obvious it will always generate a Hash).

Reply:
Thank you for the pointer to the #bury method. As you mentioned, the original proposal was rejected since it was too broad in trying to suggest #bury for Array, Struct, and Hash.

There's a follow up proposal #https://bugs.ruby-lang.org/issues/14564 which has been open since March 2018. The last comment from March 2018 says...

```
A lot of people ask for it, but I still can't see how this is a good thing to add to the language. In your personal code, sure -- maybe even in a gem -- but not the core.
```

This comment suggest that there is demand for this method but that the ruby language will not add it to the ruby implementation of a hash.